src/main/java/com/cburch/logisim/gui/main/TickCounter.java

```diff
@@ -13,172 +13,111 @@
 
 import com.cburch.logisim.circuit.Simulator;
 
-class TickCounter implements Simulator.Listener {
-  private static final int QUEUE_LENGTH = 1000;
-
-  private final long[] queueTimes;
-  private final double[] queueRates;
-  private int queueStart;
-  private int queueSize;
-  private double tickFrequency;
+import java.math.RoundingMode;
+import java.text.DecimalFormat;
+
+public class TickCounter implements Simulator.Listener {
+  private final DecimalFormat formatter;
+  private Simulator simulator;
+  private long tickCount = 0;
+  private long startTime;
+  private boolean useKiloHertz = false;
+  private double previousFrequency = 0.0;
+  private long elapsedTimeSinceLastUnitUpdate = 0;
+  static final int NANOSECONDS_UNIT_UPDATE_THRESHOLD = 500000000;
+  static final int TICKS_THRESHOLD_BEFORE_HISTORY_WEIGHT_REDUCTION = 1000000000;
+  static final int WEIGHT_REDUCTION_TICKS_COUNT = TICKS_THRESHOLD_BEFORE_HISTORY_WEIGHT_REDUCTION / 2;
```

Comment:
I recommend using the number of seconds before weight reduction. You can calculate the ticksThreshold and weightReductionTicks from the number of seconds and the computed fullTicksPerSecond. A value of around 5 seconds worked in my playing with it.

The value you have is so large that you likely never drop earlier things. At 100 ticks per second, it takes a long time to get to a billion ticks. And you really want a larger number of ticks when the clock is running fast and a smaller number when it is ticking slow.

I have sample hardware simulations that can change the speed during a run because parts of the hardware shut down (i.e. don't change state). So I can test it on that. It should change the reported speed as the simulation changes how fast it is actually running.

Reply:
It was a mistake I introduced when switching to nanoseconds.
I was supposed to be 1000.
Hopefully, it'll give a better feeling like that.